core/src/stage.cpp

```diff
@@ -219,28 +219,32 @@ void StagePrivate::sendBackward(InterfaceState&& from, const InterfaceState& to,
 	newSolution(solution);
 }
 
-void StagePrivate::spawn(InterfaceState&& state, const SolutionBasePtr& solution) {
+void StagePrivate::spawn(InterfaceState&& from, InterfaceState&& to, const SolutionBasePtr& solution) {
 	assert(prevEnds() && nextStarts());
+	
+	computeCost(from, to, *solution);
 
-	computeCost(state, state, *solution);
-
-	if (!storeSolution(solution, nullptr, nullptr))
+	if (!storeSolution(solution, &from, &to))
 		return;  // solution dropped
 
-	auto from = states_.insert(states_.end(), InterfaceState(state));  // copy
-	auto to = states_.insert(states_.end(), std::move(state));
+	auto from_it = states_.insert(states_.end(), InterfaceState(from));
+	auto to_it = states_.insert(states_.end(), std::move(to));
 
-	solution->setStartState(*from);
-	solution->setEndState(*to);
+	solution->setStartState(*from_it);
+	solution->setEndState(*to_it);
 
 	if (!solution->isFailure()) {
-		prevEnds()->add(*from);
-		nextStarts()->add(*to);
+		prevEnds()->add(*from_it);
+		nextStarts()->add(*to_it);
 	}
 
 	newSolution(solution);
 }
 
+void StagePrivate::spawn(InterfaceState&& state, const SolutionBasePtr& solution) {
+    spawn(std::move(state), std::move(state), solution); // copy start and end states
```

Comment:
You cannot move `state` twice. The first `std::move` needs to create a copy:
```suggestion
    spawn(InterfaceState(state), std::move(state), solution);
```